fe/fe-core/src/main/java/org/apache/doris/nereids/CascadesContext.java

```diff
@@ -197,4 +205,70 @@ private CascadesContext execute(Job job) {
         jobScheduler.executeJobPool(this);
         return this;
     }
+
+    public void addToTable(Table table) {
+        tables.add(table);
+    }
+
+    public void lockTableOnRead() {
+        for (Table t : tables) {
+            t.readLock();
+        }
+    }
+
+    public void releaseTableReadLock() {
+        for (Table t : tables) {
+            t.readUnlock();
+        }
+    }
+
+    /**
+     * Find all tables to lock on.
+     */
+    public void extractTables() {
```

Comment:
add UT to ensure we could lock all tables when statement has view, cte, subquery in from clause and subquery in where/having clause